### PR TITLE
lxc: Differentiate a storage device name from a storage device path

### PR DESCRIPTION
**Pull request by gabrielmougard**
closes https://github.com/canonical/lxd/issues/13238

When calling `lxc storage volume attach <pool> <volume_name> <instance_name> [device_name] [device_path]`, we consider the third argument (either [device_name] or [device_path]) to be a `device_path` only if the volume content type is `filesystem` , else this is considered a device name

### CODE REVIEW

**Comment by gabrielmougard** [comment]
@hamistao what do you think about this fix?

**Review by hamistao — CHANGES_REQUESTED** [review]

File: lxc/storage_volume.go [hunk 1]
```diff
@@ -189,26 +189,31 @@ func (c *cmdStorageVolumeAttach) Run(cmd *cobra.Command, args []string) error {
...
 	devPath := ""
 	devName := ""
 	if len(args) == 3 {
 		devName = args[1]
 	} else if len(args) == 4 {
-		// Only the path has been given to us.
-		devPath = args[3]
-		devName = args[1]
+		// Either a device name or a target path has been given to us.
+		if strings.HasPrefix(args[3], "/") || volType == "custom" {
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxc/storage_volume.go]: fixed

**Comment by gabrielmougard** [comment]
@hamistao I updated the approach

**Comment by hamistao** [comment]
Now it looks good to me! I still think it is worth it to consider only accepting paths preceded by `/` (or at least appending `/` before the path when missing before adding it to the device config), but this is not a blocker. I also think it is worth it to add a check for the server being clustered since we are now using `flagTarget`, not a blocker either because there are occurences of not including this check though the file.

**Review by hamistao — APPROVED** [review]

**Comment by tomponline** [comment]
> we consider the third argument (either [device_name] or [device_path]) to be a `device_path` only if it starts by `/` , else this is considered a device name

Would this break existing workflows that don't use `/` at the start of device paths?

If so, I don't think we can use this change as it would be a user space change.

**Comment by hamistao** [comment]
I think the current implementation accepts paths that don't use `/` at the start and just appends it if missing.

**Comment by tomponline** [comment]
> I think the current implementation accepts paths that don't use `/` at the start and just appends it if missing.

OK so we probably shouldnt change this in that case

**Comment by hamistao** [comment]
Sorry, I meant to say this PR's implementation currently works like this, so just changing the PR's description should be enough

**Comment by gabrielmougard** [comment]
@hamistao @tomponline updated PR description

File: lxc/storage_volume.go [hunk 2]
```diff
@@ -189,26 +189,49 @@ func (c *cmdStorageVolumeAttach) Run(cmd *cobra.Command, args []string) error {
...
+		vol, _, err := client.GetStoragePoolVolume(resource.name, volType, volName)
+		if err != nil {
+			return err
+		}
+
+		switch vol.ContentType {
+		case "block", "iso":
+			devName = args[3]
+		case "filesystem":
+			if !strings.HasPrefix(args[3], "/") {
```
Comment (tomponline) [inline lxc/storage_volume.go:219]: Why is this needed? If it is needed it should be in a separate commit as not related to the commit message.

Comment (hamistao) [reply lxc/storage_volume.go:219]: This is probably my [suggestion](https://github.com/canonical/lxd/pull/13258#issuecomment-2040276569), I think it makes sense to append `/` on the start of the path to better convey where the device is being mounted (showing `/something` instead of `something`). So it is indeed unrelated to the commit message.

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
Conflicts needs a rebase please

**Review by tomponline — DISMISSED** [review]

**Review by tomponline — APPROVED** [review]
